src/deploy/functions/params.ts

```diff
@@ -0,0 +1,255 @@
+
+  // A long description of the parameter's purpose and allowed values. If omitted, UX will not
+  // provide a description of the parameter
+  description?: string;
+
+  // Default value. If not provided, a param must be supplied.
+  default?: T | build.Expression<T>;
+
+  // default: false
+  immutable?: boolean;
+}
+
+export interface StringParam extends ParamBase<string> {
+  type?: "string";
+
+  // If omitted, defaults to TextInput<string>
+  input?: TextInput<string> | SelectInput<string>;
+}
+
+export interface IntParam extends ParamBase<number> {
+  type: "int";
+
+  // If omitted, defaults to TextInput<number>
+  input?: TextInput<number> | SelectInput<number>;
+}
+
+export interface TextInput<T, Extensions = {}> {
+  type?: "text";
+
+  text:
+    | Extensions
+    | {
+        example?: string;
+      };
+}
+
+interface SelectOptions<T> {
+  // Optional human-facing value for this option (e.g. "US Central (Iowa)" instead of value
+  // "us-central1")
+  label?: string;
+
+  // Actual value of the parameter if this option is selected
+  value: T;
+}
+
+export interface SelectInput<T> {
+  type?: "select";
+
+  select: Array<SelectOptions<T>>;
+}
+
+export type Param = StringParam | IntParam;
+
+/**
+ * Returns the resolved value of a user-defined Functions parameter.
+ * Functions params are defined by the output of the Functions SDK, but their value is not set until deploy-time.
+ *
+ * For most param types, we check the contents of the dotenv files first for a matching key, then interactively prompt the user.
+ * When the CLI is running in non-interactive mode or with the --force argument, it is an error for a param to be undefined in dotenvs.
+ */
+export async function handleParam(
+  param: Param,
+  projectId: string,
+  userEnvs: Record<string, string>
+): Promise<string | number | boolean> {
+  const paramName = param.param;
+
+  if (userEnvs.hasOwnProperty(paramName)) {
+    return userEnvs[paramName];
+  }
+
+  switch (param.type) {
+    case "string":
+      return promptStringParam(param);
+    case "int":
+      return promptIntParam(param);
+    default:
+      throw new FirebaseError(
+        "Build specified parameter " + param.param + " with unsupported type"
+      );
+  }
+}
+
+async function promptStringParam(param: StringParam): Promise<string> {
+  if (!param.input) {
+    if (param.default) {
+      return resolveString(param.default, {});
+    }
+    throw new FirebaseError(
+      "Build specified string parameter " + param.param + " without any input form or default value"
+    );
+  }
+
+  switch (param.input.type) {
+    case "select":
+      throw new FirebaseError(
+        "Build specified string parameter " + param.param + " with unsupported input type 'select'"
+      );
+    case "text":
+    default:
+      let prompt = `Enter a value for ${param.label || param.param}:`;
+      if (param.description) {
+        prompt += ` \n(${param.description})`;
+      }
+      return await promptOnce({
```

Comment:
Doesn't this mean the user is going to get prompted on ever deploy? We need to write the params back to the appropriate .env file.

Reply:
Made a note of that, but I think this PR is large enough already.